### PR TITLE
Custom parser(s) support and more encapsulation

### PR DESCRIPTION
**Pull request by delvedor**
With this pr we add the support of custom parsers based on the content type.
The `application/json` is always handled by fastify, in this way we have a fast case for the most common use case and we don't add overhead.

Usage:
```js
fastify.addParseStrategy('application/jsoff', function (req, done) {
  myCustomParser(req, function (err, body) {
    done(err || body)
  })
})
```
Feedbacks?

### CODE REVIEW

**Comment by mcollina** [comment]
how does this work with plugins? Can two plugins define two different way of parsing the same content-type?

**Comment by mcollina** [comment]
I would call the method `addContentTypeParser`, it's lengthy but explicative.

**Comment by delvedor** [comment]
> how does this work with plugins? Can two plugins define two different way of parsing the same content-type?

I didn't try, but I'm pretty sure that this implementation does not allow encapsulation, is very similar to the hooks one.

**Comment by delvedor** [comment]
Now the `ContentTypeParser` is fully encapsulated! 🎉 

Regarding the hooks, `preHandler` is encapsulated but I'm having some trouble to encapsulate `onRequest` and `preRouting` as well, the problem is [here](https://github.com/fastify/fastify/blob/5d359bdb43def6da90fa4ed30d8d7ad577c6eeee/fastify.js#L101) and [here](https://github.com/fastify/fastify/blob/5d359bdb43def6da90fa4ed30d8d7ad577c6eeee/fastify.js#L117), because obviously I'm not using the correct instance of fastify. At the moment I don't know how to fix this.

Ideas? :)

ps I didn't tested `onClose` yet, but I'm pretty sure it will have the same issue.

File: test/custom-parser.test.js [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+  fastify.post('/', (req, reply) => {
+    reply.send(req.body)
+  })
+
+  fastify.contentTypeParser.add('application/jsoff', function (req, done) {
+    jsonParser(req, function (err, body) {
+      if (err) return done(err)
+      done(body)
+    })
+  })
```
Comment (mcollina) [inline test/custom-parser.test.js]: I think we should should wrap `contentTypeParser.add` into a single function, and hide the `contentTypeParser` property.

File: test/hooks.test.js [hunk 2]
```diff
@@ -3,12 +3,13 @@
...
 const test = t.test
 const request = require('request')
+const Fastify = require('..')
 const fastify = require('..')()
 
 test('hooks - add preHandler', t => {
   t.plan(1)
   try {
-    fastify.addHook('preHandler', function (request, reply, next) {
+    fastify.hooks.add('preHandler', function (request, reply, next) {
```
Comment (mcollina) [inline test/hooks.test.js]: why can't `addHook` just call `fastify[kHooks].add` internally?

File: fastify.js [hunk 3]
```diff
@@ -317,6 +319,26 @@ function build (options) {
...
+  function buildHooks (h) {
+    function _Hooks () {}
+    _Hooks.prototype = new Hooks()
+    const H = new _Hooks()
+    H.onRequest = Object.create(h.onRequest)
+    H.preRouting = Object.create(h.preRouting)
+    H.preHandler = Object.create(h.preHandler)
+    H.onClose = Object.create(h.onClose)
+    return H
+  }
```
Comment (mcollina) [inline fastify.js]: this should probably be in the `Hooks` file.

**Comment by mcollina** [comment]
I think we should disable setting _those_ hooks in encapsulated plugins then, as I do not see how it is possible (we can only set them in top-level plugins) to support encapsulation. It's better to throw and avoid confusion for the users.

we already have [here](https://github.com/fastify/fastify/blob/6e0dcbc43a768154a196d37e86c1313e5cb5dab2/lib/tier-node.js#L88-L98) the `preHandler` hook, that can be fully encapsulated, if I understand correctly.

BTW, should we have a `postHandler`?

File: fastify.js (hunk 3 above)
Comment (delvedor) [inline fastify.js]: The same for `buildContentTypeParser` I guess (maybe Reply and Request too?)

Comment (mcollina) [inline fastify.js]: yes!

**Comment by delvedor** [comment]
Let me understand , your suggestion is to force the users declare the hook in top level?

Regarding the `preHandler` yes, it can be fully encapsulated by adding it to the node object as we are doing with Reply and Request.

> BTW, should we have a `postHandler`?

I don't know, what are the benefits?
I think add it would mean change heavily Reply .

**Comment by delvedor** [comment]
Little recap:
- `ContentTypeParser` is fully encapsulated
- `preHandler` hook is encapsulated, while `onRequest`, `preRouting` and `onClose` are not
- the hooks runners are wrapped with a `setImmediate`, this give us at least 2k req/sec, but we must test how it behave with CPU bound operations
- some little refactor in *tier-node*

**Pull request merged**